src/main/java/me/karakelley/http/exceptions/PublicDirectoryMissingException.java

```diff
@@ -0,0 +1,7 @@
+package me.karakelley.http.exceptions;
+
+public class PublicDirectoryMissingException extends RuntimeException{
```

Comment:
What do you get by extending `RuntimeException` instead of just `Exception`? Is there a reason some extend one while others extend the other?

Reply:
If I'm honest, I didn't have a good reason but TIL `Exception` is checked while `RuntimeException` is unchecked. At this point, I don't want/need these particular exceptions to be checked so I changed them all to extend `RuntimeException`

Reply:
So is it true that by making them unchecked, clients can call methods that potentially throw these exceptions without needing a `try/catch` block? In which case when one is thrown, it'll bubble up to something higher... how far would that go in your server today?

I've been fighting (and losing) a battle over exceptions in Scala with some client colleagues. There are some patterns I've wanted to try in that language to avoid exceptions as much as possible, but I recognize some downsides... And of course this is Java, which is different...

I dunno. @aaronlahey do you have any thoughts on this?

Reply:
I have a lot of thoughts on this but no nice clean rule of thumb...

Having a bunch of named exceptions was a pattern I used pretty successfully at a client. It allows for a "poor man's pattern matching."

```
try {
  ...
} catch (FooWentWrong | BarWentWrong) {
  ...
} catch (BazWentWrong) {
  ...
} catch (QuuxWent Wrong) {
  ...
}
```

I think exceptions tend to make things more complicated than people realize...

1. They circumvent the control flow
2. You don't really know what's going to be thrown and for what reasons (unless they're checked exceptions but then you're polluting everything with `throws Exception`).
3. You have to be extra careful when you have resources you need to free (by using `try`/`finally` or `try(...)`.

But, they *can* be less complicated than result types in a language like java and they can be cheaper if you're doing really low-level things (you can avoid allocating an object if you just need to return an `int` or something).